### DISCUSSION TITLE
Battery Telemetry (RXBT) for PWM Receivers

### DISCUSSION THREAD

**Discussion by derFrickler** [ExpressLRS/ExpressLRS#1464] [Ideas]
With the new available PWM Receivers, more and more people will use ELRS on planes/wings or other vehicles without FC.

It would be really great if the battery voltage telemetry (RXBT) would be available on PWM receivers. 

Some of them already have the ADC/A0 pin of the esp8266 exposed on their layouts - like the matek elrs-r24-p. For future layouts it would be great to already add a voltage divider to the A0 to the pcb design to allow at least up to 6S input. 

It would be really great to have some code implemented to "detect" the number of connected lipo cells and divide the reported voltage by the number of cells, similar to the "report_cell_voltage" option on betaflight. This makes the TX side warnings agnostic from the number of connected battery cells. https://github.com/betaflight/betaflight/blob/master/src/main/sensors/battery.c#L205

I already tried something similar in my fork of CaptnBrys CRServo F code which works fine for me, but it would be great to have it in the new receivers already. https://github.com/derFrickler/CRServoF/blob/master/src/main.cpp#L140

Thanks a lot for the great work on ExpressLRS

**Comment by danxdz** [comment]
Thanks i talked about this, last weekend.. this is a must have for everything that dont use fc..

**Comment by CapnBry** [comment]
https://github.com/ExpressLRS/ExpressLRS/pull/1466

The problem with report_cell_voltage is that CRSF protocol sucks donkey spit because there's only one decimal place for VBAT precision (but supports up to 6553.5V because of course TBS always has the best ideas). If you want two digits of precision you're going to have to take that request up the chain to TBS and everyone else everywhere that displays a VBAT value (or use a lower ANALOG_VBAT_SCALE).

You also can't reliably lock on to the number of cells from just the single first measurement due to the first readings possibly being too low due to caps charging or other ramping. What I have done in the past is keep reading for a while and calculate the cellcount, and when it gets the same value X times in a row (I used to use 100 but that's a lot for a low telemetry rate), it locks in the cell count at that time. I may consider that, but the issue is that again there's only 1 decimal place in the CRSF voltage so the cell voltage numbers are really rough.

**Comment by derFrickler** [comment]
Thanks CantainBry
Sure, doing multiple readings before would be the better way, i did that for other projects. 
In your CRServoF its works fine without that for me so far, as the code needs some time anyway to statup up before doing the first voltage reading and you already do some or smoothing/averaging already.
Cell Voltage with one Digit only would be perfectly fine for me (and most others i think). always only use one digit in the OSD and OTX screen/warning as well. Its more about to know if we are talking 4.1V or 3.3V on a planes battery.

**Comment by elvenpure** [comment]
@derFrickler are you saying it is possible to use the adc pad to receive battery voltage via telemetry? I don't mind if it is not super accurate, but would save many people buying a seperate crsf to pwm and receiver. I am litterally about to order some now haha, and would love a better/cheaper solution! Cheers.

**Comment by derFrickler** [comment]
Yes, technically its possible, just the code to do that needs to be written.

**Comment by wernight** [comment]
ELRS-R24-V is the only PWM ELRS receiver that I found so far that seem to report the battery voltage without a FC. The rest seem to cap at around 10V input voltage, so I'm not going to connect even 3 cell battery to it.

**Reply by CapnBry** [reply]
Most PWM RX have a separate VBAT pad and do not use the input voltage for VBAT.

**Reply by wernight** [reply]
@CapnBry That's very good to know.

These are the popular and cheap ELRS PWM receiver I found:

 - [HappyModel EPW5](https://www.aliexpress.com/item/1005003749317689.html) seems to have no such VBAT
 - [Mateksys ELRS-R24-P](http://www.mateksys.com/?portfolio=elrs-r24-p) says `Input voltage: 4~9V DC @ “+” pad` so I guess again no VBAT for 3+ cells
 - [Betafpv Micro Receiver](https://betafpv.com/products/elrs-micro-receiver) says `5V~9V DC @ "+" pad` (idem)

Am I mistaken here or is there some other cheap receiver that can do battery voltage telemetry for PWM?

**Reply by commanderguy3001** [reply]
I think what you are referencing is the input voltage to the receiver, with which it is powered, not the voltage sensing.
In general any esp82xx chip can only measure from 0V to 1V on the analog pins, so you will need a voltage divider anyway, which I don't think any receiver has built-in yet, so you will have to build out of two specific resistors in series between ground and your battery voltage, and the point between the two will be your measurement point which will be a scaled voltage based on your battery voltage, which the esp can read, and is then multiplied back by the scaling factor to get the actual voltage.
Depending on what resistors you use you can basically get whatever voltage range you want, though you might have to change the software a little to use a different scaling factor, I think the default only works up to 6S voltage.

**Comment by CapnBry** [comment]
Commanderguy is right, it does not measure the receiver supply voltage, vbat must be fed in on a separate place. On the Matek, there is a pad labeled ADC on the back of the receiver, but the max voltage there needs to be at most 1V, thus requiring the voltage divider cmdr mentioned. The values I can't remember offhand and I am in vacation and it is difficult to look up. If you're going to try that then post back here and I can provide the two resistor values needed.

The happy model receiver does not have this input for sure, and I don't know about the beta FPV.

**Reply by wernight** [reply]
Yup I see the ADC on the back. So basically if you have plan to have at most 3 cells LiHV, you'd want the max voltage to be 4.35 x 1.1 x 3 = 14.35 V (keeping some safety margin), so you'd want 2 resistors like:

  - R2 = 100 kΩ
  - R1 = 100 x (14.35 - 1) = 1335 kΩ ≈ 1.3 MΩ

That sounds great. Then I guess the scaling would be done on the OpenTX radio or can it be done from the Rx?

btw. this would be awesome on some Wiki. I know at least how the voltage divider work but most people probably would want a more step by step guide.

**Reply by commanderguy3001** [reply]
first: you don't want _that_ low of a resistance, as there still needs to happen some current flow, so rather use 20k/260k.
second: there are specific scaling values already predefined somewhere in the software, which you cannot change easily, so you gotta use the correct resistances for that scalar unless you wanna change the software
third: I didn't do the maths on your calculation, but in general I think you understood it
fourth: the scaling is done only on the receiver, as opentx/edgetx just receives the voltage as a voltage, and just displays it, and doesn't do any maths on it (basically)
fifth: no wiki yet, cause afaik it's not yet in the current release (v2.5.0)

**Comment by CapnBry** [comment]
It is on the ballpark of 24k and 1k. It isn't in the documentation because it isn't released. The documentation is only for the released software. You can't scale it and open text because it's not flexible enough to do that kind of math you'd have to build the image with the proper scale defined in user defines if you do not use the values that I tell you, which I told you I will look up next week

**Reply by wernight** [reply]
So by default it assumes a 1:25 scale? Which should support up to 6S LiPo by my calculation (which is good). By "not released", this 1:25 scaling telemetry from ADC is not part of ELRS 2.5 but likely to be part of ELRS 2.6?

**Reply by commanderguy3001** [reply]
there probably won't be a version 2.6 I reckon, as version 3.0 is pretty close already, so I guess it will be part of that.
Here is the github PR regarding vbat telemetry: [#1466](https://github.com/ExpressLRS/ExpressLRS/pull/1466)

**Comment by xsnoopy** [comment]
Hi everyone, just found this discussion. I am using a Radiomaster TX16S with TBS and ESLRS. I converted all my Planes,... which fly without an FC to Flysky receiver with a battery sensor. With ESLRS 3.0 I'm almost ready to switch everything from Flysky to ESLRE if I could get a good way of reading the battery voltage. Perfect would be if I could also see the used mAh, but I can't even see that currently.

Reading this discussion and seeing that we have only .1 digit in the vbat reading. Would it be possible to day the readings are all 10x ? Like a voltage of 11.5V would read 115.0V it could be devided in Radio by 10 and you effectively have 1 digit more. 

I received with my flysky set up only the total battery voltage. Since I have models flying from 2s to 6s I put a hand full of logical switches together which give me:

1. A delay in the first voltage reading until the battery voltage is stabilized (simple delay timer of a few seconds)

2. Calculate how many cells I have loaded. Important since some of my models I fly with batteries from 3 to 6s and the characteristics change dramatically on the throttle stick :)

3. Calculate the cell voltage and give me verbal feedback of the cell voltage drops below 3.7V 

The flysky battery sensor is probably just a resistor array with a voltage devider maybe some diods for protection,... Wouldn't it be an easy thing to come up with some simple open hardware circuit which would do exactly that and manufacturer could jump on that and build them eslrs compatible so to say?

**Comment by xsnoopy** [comment]
Hey, I see now more and more receiver with Vbat pads. For example the Matek ExpressLRS/ELRS 2.4GHZ PWM Receiver. It should be able to read up to 34V on the vbat pad. I have the receiver at home, just need to test it.

**Comment by bd4** [comment]
The RadioMaster ER4 and ER6 (probably ER8 too?) have a V-EXT port for battery voltage, according to the manuals. On ER6 it's an extra JST style connection port; on the ER4 it needs to be soldered (and possibly unwrapped and rewrapped? Will look more closely at mine tomorrow). Note that the ER5C does not appear to have it. My understanding is that the VBat is for input power to the receiver, which in many cases comes in on the ESC/throttle port and is not needed.
